CMakeLists.txt

```diff
@@ -1,23 +1,111 @@
-cmake_minimum_required(VERSION 3.0)
+cmake_minimum_required(VERSION 3.3) # support visibility settings
 
-list(APPEND CMAKE_MODULE_PATH
-       "${CMAKE_CURRENT_SOURCE_DIR}/cmake"
-       "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules")
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake/Modules")
 
-project(thread-pool-cpp)
-add_definitions(-std=c++14 -Wall -Werror -O3)
+project(thread-pool-cpp VERSION 1.1.0)
 
+option(THREAD_POOL_CPP_BUILD_TESTS "Build tests." OFF)
+option(THREAD_POOL_CPP_BUILD_BENCHMARKS "Build benchmarks." OFF)
 
-# Get all include files
+if(THREAD_POOL_CPP_BUILD_TESTS)
+  include(CTest)
+endif()
+
+# Variable used in: benchmark, tests
 set(THREAD_POOL_CPP_INC_DIR "${CMAKE_CURRENT_SOURCE_DIR}/include/")
 include_directories("${THREAD_POOL_CPP_INC_DIR}")
-file(GLOB_RECURSE INSTALL_FILES_LIST "${THREAD_POOL_CPP_INC_DIR}/*")
 
-add_subdirectory(tests)
-add_subdirectory(benchmark)
+# If true C++11 thread_local support exists, we will use it:
+include(thread_pool_has_thread_local_storage)
+thread_pool_has_thread_local_storage(THREAD_POOL_HAS_THREAD_LOCAL_STORAGE)
+message("THREAD_POOL_HAS_THREAD_LOCAL_STORAGE: ${THREAD_POOL_HAS_THREAD_LOCAL_STORAGE}")
+
+if(NOT THREAD_POOL_HAS_THREAD_LOCAL_STORAGE)
+  # Else, we will check for backups
+  include(thread_pool_has_thread_storage)
+  thread_pool_has_thread_storage(THREAD_POOL_HAS_THREAD_STORAGE)
+  message("THREAD_POOL_HAS_THREAD_STORAGE: ${THREAD_POOL_HAS_THREAD_STORAGE}")
+  if(NOT THREAD_POOL_HAS_THREAD_STORAGE)
+    message(FATAL_ERROR "Compiler does not support: thread_local, __thread, or declspec(thread)")
+  endif()
+endif()
+
+# http://stackoverflow.com/a/11583676
+add_library(thread-pool-cpp INTERFACE)
+
+option(THREAD_POOL_CPP_HAS_CPP14 "Has true C++14 support." ON)
+
+if(NOT MSVC)
```

Comment:
I think something like this might work (and is a little cleaner):
```
if(THREAD_POOL_HAS_CPP14)
  set(CMAKE_CXX_STANDARD 14)
else()
  set(CMAKE_CXX_STANDARD 11)
endif()

target_compile_features(
  thread-pool-cpp
  INTERFACE
  cxx_auto_type
  cxx_nullptr
  cxx_lambdas
  cxx_lambda_init_captures
  )
```